File: Array-Strig/ProductOfArrayExceptSelf.java
public class ProductOfArrayExceptSelf{
  public static void main(String[] args){
    Solution s = new Solution();
    int[] result = s.productExceptSelf(new int[]{1,2,3,4});
    for(int i = 0; i < result.length; ++i)
      System.out.print(result[i] + " ");
    System.out.print("\n");
    result = s.productExceptSelf(new int[]{-1,1,0,-3,3});
    for(int i = 0; i < result.length; ++i)
      System.out.print(result[i] + " ");
    System.out.print("\n");
  }
}

class Solution{
  public int[] productExceptSelf(int[] nums){
    int length = nums.length;
    int[] result = new int[length];
    int product = 1;
    for(int i = 0; i < length; ++i){
      result[i] = product;
      product *= nums[i];
    }
    product = 1;
    for(int i = length-1; i >=0; --i){
      result[i] *= product;
      product *= nums[i];
    }
    return result;
  }
}

File: LC 75/ProductOfArrayExceptSelf.go
package main

import "fmt"

func productExceptSelf(nums []int) []int {
	result := make([]int, len(nums))
	result[0] = 1
	for i := 1; i < len(nums); i++ {
		result[i] = result[i-1] * nums[i-1]
	}
	temp := 1
	for i := len(nums) - 1; i >= 0; i-- {
		result[i] *= temp
		temp *= nums[i]
	}
	return result
}

func main() {
	nums := []int{1, 2, 3, 4}
	fmt.Println(productExceptSelf(nums))
}
